source/opt/ir_loader.cpp

```diff
@@ -131,40 +181,53 @@ bool IrLoader::AddInstruction(const spv_parsed_instruction_t* inst) {
         return false;
       }
     } else {
-      if (block_ == nullptr) {  // Inside function but outside blocks
-        if (opcode != SpvOpFunctionParameter) {
-          Errorf(consumer_, src, loc,
-                 "Non-OpFunctionParameter (opcode: %d) found inside "
-                 "function but outside basic block",
-                 opcode);
-          return false;
-        }
-        function_->AddParameter(std::move(spv_inst));
-      } else {
-        if (opcode == SpvOpExtInst &&
-            spvExtInstIsDebugInfo(inst->ext_inst_type)) {
-          const uint32_t ext_inst_index = inst->words[4];
-          if (inst->ext_inst_type == SPV_EXT_INST_TYPE_OPENCL_DEBUGINFO_100) {
-            const OpenCLDebugInfo100Instructions ext_inst_key =
-                OpenCLDebugInfo100Instructions(ext_inst_index);
-            if (ext_inst_key != OpenCLDebugInfo100DebugScope &&
-                ext_inst_key != OpenCLDebugInfo100DebugNoScope &&
-                ext_inst_key != OpenCLDebugInfo100DebugDeclare &&
-                ext_inst_key != OpenCLDebugInfo100DebugValue) {
+      if (opcode == SpvOpLoopMerge || opcode == SpvOpSelectionMerge)
+        last_dbg_scope_ = DebugScope(0, 0);
+      if (last_dbg_scope_.GetLexicalScope())
+        spv_inst->SetDebugScope(last_dbg_scope_);
+      if (opcode == SpvOpExtInst &&
+          spvExtInstIsDebugInfo(inst->ext_inst_type)) {
+        const uint32_t ext_inst_index = inst->words[kExtInstSetIndex];
+        if (inst->ext_inst_type == SPV_EXT_INST_TYPE_OPENCL_DEBUGINFO_100) {
+          const OpenCLDebugInfo100Instructions ext_inst_key =
+              OpenCLDebugInfo100Instructions(ext_inst_index);
+          switch (ext_inst_key) {
+            case OpenCLDebugInfo100DebugDeclare: {
+              const uint32_t local_var_id = inst->words[kLocalVariableIdIndex];
+              module_->AddDebugLocalVariableInfo(local_var_id,
+                                                 std::move(spv_inst));
+              break;
+            }
+            case OpenCLDebugInfo100DebugValue: {
+              const uint32_t value_id = inst->words[kLocalVariableIdIndex];
+              module_->AddDebugLocalVariableInfo(value_id, std::move(spv_inst));
```

Comment:
For debug value, you have to leave that in the body of the function at the position that it appears.  That is because it says "starting here the value of the variable is obtained this way".  If there are multiple debug values for the same variable, how do you know which one to applies?  

I think that debug declare should be in the function body as well.  We  can discuss this again.  If needed, we can build a map from a variable to it debug declare instruction as an analysis.